File: Stack/ValidParenthesis.c
#include "stdio.h"
#include "stdlib.h"
#include "string.h"

struct Stack{
    int size;
    int top;
    char *s;
};

void push(struct Stack *st, char val){
    st->top++;
    st->s[st->top] = val;
}

int isEmpty(struct Stack st){
    return st.top==-1;
}

void pop(struct Stack *st){
    st->top--;
}

int isBalanced(char *exp){
    struct Stack st;
    st.size = strlen(exp);
    st.top = -1;
    st.s = (char *)malloc(st.size * sizeof(char));
    for (int i=0;exp[i]!='\0';i++){
        if (exp[i]=='(') push(&st, exp[i]);
        else if (exp[i]==')'){
            if (isEmpty(st)) return 0;
            else pop(&st);
        }
    }
    return isEmpty(st);
}

int main(){
    char *exp = "((a+b)*(c-d)))";
    printf("%d",isBalanced(exp));
}

File: Stack/ValidParenthesis.py

class Stack:
    def __init__(self):
        self.top = -1
        self.size = 0
        self.s = []
    
    def push(self, val: str) -> None:
        self.top += 1
        self.s[self.top] = val
    
    def pop(self) -> str:
        self.top -= 1
        
    def isEmpty(self) -> bool:
        return self.top==-1
        
def isBalanced(s: str) -> bool:
    st = Stack()
    st.size = len(s)
    st.s = [''] * st.size
    for c in s:
        if c=='(':
            st.push('(')
        elif c==')':
            if st.isEmpty(): return False
            else: st.pop()
    return st.isEmpty()

s = "((a*b))*((c*d))"
print(isBalanced(s))

